modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionDemander.java

```diff
@@ -1837,6 +1935,87 @@ public boolean compatibleWith(GridDhtPreloaderAssignments newAssignments) {
             return true;
         }
 
+        /**
+         * Callback when getting entry from supplier.
+         *
+         * @param p Partition.
+         * @param node Supplier.
+         */
+        private void onReceivedKeys(int p, ClusterNode node) {
+            receivedKeys.incrementAndGet();
+
+            boolean hist = assignments.get(node).partitions().hasHistorical(p);
+            (hist ? histReceivedKeys : fullReceivedKeys).get(node.id()).increment();
+        }
+
+        /**
+         * Callback when getting size of received entries in bytes.
+         *
+         * @param p Partition.
+         * @param bytes Byte count.
+         * @param node Supplier.
+         */
+        private void onReceivedBytes(int p, long bytes, ClusterNode node) {
+            boolean hist = assignments.get(node).partitions().hasHistorical(p);
+            (hist ? histReceivedBytes : fullReceivedBytes).get(node.id()).add(bytes);
+        }
+
+        /**
+         * Checks if rebalance chain has completed.
+         *
+         * @return {@code true} if rebalance chain completed.
+         */
+        private boolean isChainFinish() {
+            RebalanceFuture fut = this;
+
+            while (nonNull(fut)) {
+                if (fut.isInitial() || !fut.isDone() || !fut.result())
+                    return false;
+                else
+                    fut = fut.next;
+            }
+
+            return true;
+        }
```

Comment:
It still doesn't look correct to me.
Consider a rebalance chain g1 -> g2 ->g3 for topVer=N, g2 is currently rebalancing.
Topology is changed to N + 1, g3 is incompatible and cancelled.
After g2 has been finished rebalancing a chain will not be completed because g3.rebFut.result=false, but it should.
Also, no need to copy reference to _this_.

The correct implementation for me looks something like:

```
private boolean isChainFinished() {
        assert isDone() : this;

        if (isInitial() || !result())
            return false;

        RebalanceFuture fut = next;

        while (fut != null) {
            if (fut.isInitial() || !fut.isDone())
                return false;
            else
                fut = fut.next;
        }

        return true;
}
```